app/javascript/controllers/map_controller.js

```diff
@@ -0,0 +1,96 @@
+import { Controller } from "stimulus"
+import consumer from "../channels/consumer"
+
+export default class extends Controller {
+  static targets = ["image"]
+
+  connect() {
+    this.mapId = this.element.dataset.mapId
+    this.channel = consumer.subscriptions.create({
+      channel: "MapChannel",
+      id: this.mapId
+    }, {
+      received: this.cableReceived.bind(this)
+    })
+  }
+
+  moveMap(event) {
+    const speedFactor = 2
+    var mousedown = {
+      x: event.pageX,
+      y: event.pageY
+    }
+    var backgroundPosition = {
+      x: Number.parseInt(event.target.style.backgroundPosition.split(" ")[0]),
+      y: Number.parseInt(event.target.style.backgroundPosition.split(" ")[1])
+    }
+    var handleMove = (event) => {
+      var mousepos = {
+        x: event.pageX,
+        y: event.pageY
+      }
+      if (mousedown != mousepos) {
+        var moveby = {
+          x: (mousepos.x - mousedown.x) * speedFactor,
+          y: (mousepos.y - mousedown.y) * speedFactor
+        }
+        this.channel.perform(
+          "move_map",
+          {
+            map_id: this.mapId,
+            x: backgroundPosition.x + moveby.x,
+            y: backgroundPosition.y + moveby.y
+          }
+        )
+      }
+    }
+    document.addEventListener("mouseup", function () {
+      document.removeEventListener("mousemove", handleMove)
+    })
+    document.addEventListener("mousemove", handleMove)
+  }
+
+  setMapPosition(x, y) {
+    this.imageTarget.style.backgroundPosition = `${x}px ${y}px`
+  }
+
+  zoom(factor) {
+    var backgroundSize = {
+      width: Number.parseInt(this.imageTarget.style.backgroundSize.split(" ")[0]),
+      height: Number.parseInt(this.imageTarget.style.backgroundSize.split(" ")[1])
+    }
+    this.channel.perform(
+      "set_size",
+      {
+        map_id: this.mapId,
+        width: Math.round(backgroundSize.width * factor),
+        height: Math.round(backgroundSize.height * factor),
+      }
+    )
+  }
+
+  zoomIn(event) {
+    event.preventDefault()
+    this.zoom(1.25)
+  }
+
+  zoomOut(event) {
+    event.preventDefault()
+    this.zoom(0.75)
+  }
+
+  setMapSize(width, height) {
+    this.imageTarget.style.backgroundSize = `${width}px ${height}px`
+  }
+
+  cableReceived(data) {
+    switch (data.operation) {
+      case "move":
+        this.setMapPosition(data.x, data.y)
+        break
+      case "zoom":
+        this.setMapSize(data.width, data.height)
+        break
+    }
```

Comment:
There's some jittering which I think comes from additional move events happening before the socket returns and updates the location. To get around this with the tokens I:

- [Set a `beingDragged` data value to true](https://github.com/jdbann/table/pull/7/files#diff-4aa6a812af14d655f6026b2bfeedaffaR26)
- [Updated the token's location as part of the move events](https://github.com/jdbann/table/pull/7/files#diff-4aa6a812af14d655f6026b2bfeedaffaR28)
- [Ignored cable events if `beginDragged` was true](https://github.com/jdbann/table/pull/7/files#diff-4aa6a812af14d655f6026b2bfeedaffaR67-R69)
- [Set `beingDragged` to false on mouseup](https://github.com/jdbann/table/pull/7/files#diff-4aa6a812af14d655f6026b2bfeedaffaR32)

Reply:
Good call, I had a version with less jittering because I was setting it client size, and then I changed it right before opening the PR so I didn't notice how bad the jittering was. In 44f6aa1 I've changed it to use this approach and the performance is much improved.